graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphQLService.java

```diff
@@ -0,0 +1,202 @@
+        final String operationName = queryString.get("operationName");
+        return execute(executionInput(query, ctx, variables, operationName));
+    }
+
+    @Override
+    protected HttpResponse doPost(ServiceRequestContext ctx, HttpRequest request) {
+        final MediaType contentType = request.contentType();
+        if (contentType == null) {
+            return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE,
+                                   MediaType.PLAIN_TEXT,
+                                   "Could not process GraphQL request");
+        }
+        if (MediaType.isJson(contentType)) {
+            return HttpResponse.from(request.aggregate().handle((req, thrown) -> {
+                if (thrown != null) {
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String body = req.contentUtf8();
+                if (Strings.isNullOrEmpty(body)) {
+                    return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT, "Body is required");
+                }
+                final Map<String, Object> requestMap = parseJsonString(body);
+                final String query = (String) requestMap.getOrDefault("query", "");
+                return execute(executionInput(query, ctx,
+                                              toVariableMap(requestMap.get("variables")),
+                                              (String) requestMap.get("operationName")));
+            }));
+        }
+        if (contentType.is(MediaType.GRAPHQL)) {
+            return HttpResponse.from(request.aggregate().handle((req, thrown) -> {
+                if (thrown != null) {
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String query = req.contentUtf8();
+                return execute(executionInput(query, ctx, null, null));
+            }));
+        }
+        return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE,
+                               MediaType.PLAIN_TEXT,
+                               "Could not process GraphQL request");
+    }
+
+    private ExecutionInput executionInput(String query, ServiceRequestContext ctx,
+                                          @Nullable Map<String, Object> variables,
+                                          @Nullable String operationName) {
+        final Builder builder = ExecutionInput.newExecutionInput(query);
+        if (variables != null && !variables.isEmpty()) {
+            builder.variables(variables);
+        }
+        if (operationName != null) {
+            builder.operationName(operationName);
+        }
+        if (dataLoaderRegistry != null) {
+            builder.dataLoaderRegistry(dataLoaderRegistry);
+        }
+        return builder.context(ctx).build();
+    }
+
+    private HttpResponse execute(ExecutionInput input) {
+        return HttpResponse.from(graphQL.executeAsync(input).handle((executionResult, cause) -> {
```

Comment:
How about adding an `useBlockingTaskExecutor(boolean)` to `GraphQLServiceBuilder` so that `graphQL.executeAsync(input)` could be executed in a blocking task executor as we did in `GrpcServicBuilder`?
https://github.com/line/armeria/blob/480321970111ac6249775199fedadcc71b862e84/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java#L402

Reply:
Good idea 👍